llama-index-core/llama_index/core/llms/function_calling.py

```diff
@@ -0,0 +1,165 @@
+from typing import (
+    Any,
+    List,
+    Optional,
+    Union,
+    TYPE_CHECKING,
+)
+import asyncio
+
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+)
+from llama_index.core.llms.llm import LLM
+
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+)
+from llama_index.core.llms.llm import ToolSelection
+
+if TYPE_CHECKING:
+    from llama_index.core.chat_engine.types import AgentChatResponse
+    from llama_index.core.tools.types import BaseTool
+
+
+class FunctionCallingLLM(LLM):
+    """
+    Function calling LLMs are LLMs that support function calling.
+    They support an expanded range of capabilities.
+
+    """
+
+    def chat_with_tools(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Predict and call the tool."""
+        raise NotImplementedError("predict_tool is not supported by default.")
```

Comment:
should this be `chat_with_tools` rather than `predict_tool`?

Reply:
yes oops